### PR TITLE
Planner cleanup

### PR DESCRIPTION
**Pull request by JordySehn**
Merging planner cleanup onto main - tested extensively on warthog. Includes high level planner behaviours, improved stability, and more reliable obstacle avoidance.

### CODE REVIEW

File: main/src/vtr_path_planning/src/mpc/speed_scheduler.cpp [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+
+double ScheduleSpeed(const std::vector<double>& disc_path_curvature_xy, const std::vector<double>& disc_path_curvature_xz_yz, double VF, int curr_sid, double planar_curv_weight, double profile_curv_weight, double eop_weight, double horizon_step_size)
+{
+
+    // Experimental Speed Scheduler:
+    // Takes in the desired forward_velocity and the pre-processed global path and reduces the set speed based on a range of tunable factors:
+    // 1. Planar Curvature
+    // 2. Profile Curvature
+    // 3. End of Path
+    // 4. Corridor Width (TODO)
```
Comment (a-krawciw) [inline main/src/vtr_path_planning/src/mpc/speed_scheduler.cpp:33]: Are these still TODO

File: main/src/vtr_pose_graph/include/vtr_pose_graph/path/path.hpp [hunk 2]
```diff
@@ -192,6 +195,12 @@ double Path<GraphT>::dist(unsigned seq_id) const {
   return distances_[seq_id];
 }
 
+template <class GraphT>
+int Path<GraphT>::query_terrain_type(unsigned seq_id) const {
+  int terrain_type = graph_->at(VertexId(0,seq_id))->GetTerrainType();
```
Comment (a-krawciw) [inline main/src/vtr_pose_graph/include/vtr_pose_graph/path/path.hpp:200]: Hard coding major id 0 is not ideal. Other teach passes can have greater than 0.

File: rviz/honeycomb.rviz
Comment (a-krawciw) [inline rviz/honeycomb.rviz:1]: Should other sensor rviz be updated too?

File: main/src/vtr_tactic/src/tactic.cpp [hunk 3]
```diff
@@ -165,7 +165,7 @@ bool Tactic::routeCompleted() const {
   auto lock = chain_->guard();
   const auto translation = chain_->T_leaf_trunk().r_ab_inb().norm();
 
-  if (chain_->trunkSequenceId() < (chain_->sequence().size() - 1)) {
+  if (chain_->trunkSequenceId() < (chain_->sequence().size() - 2)) {
```
Comment (a-krawciw) [inline main/src/vtr_tactic/src/tactic.cpp:168]: Why the change?

File: main/src/vtr_path_planning/src/mpc/speed_scheduler.cpp [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+
+    }
+    avg_curvature_xy = avg_curvature_xy / horizon_steps;
+    avg_curvature_xz_yz = avg_curvature_xz_yz / horizon_steps;
+    CLOG(INFO, "mpc.speed_scheduler") << "THE AVERAGE PLANAR CURVATURE IS:  " << avg_curvature_xy;
+    CLOG(INFO, "mpc.speed_scheduler") << "THE AVERAGE PROFILE CURVATURE IS:  " << avg_curvature_xz_yz;
+
+    // handle forward/referse case and calculate a candidate VF speed for each of our scheduler modules (XY curvature, XZ curvature, End of Path etc)
+
+    VF_EOP = std::max(0.5, VF / (1 + (end_of_path * end_of_path * eop_weight)));
```
Comment (a-krawciw) [inline main/src/vtr_path_planning/src/mpc/speed_scheduler.cpp:69]: Is this the min speed? Should it be a parameter?

File: main/src/vtr_path_planning/src/mpc/mpc_path_planner.cpp [hunk 5]
```diff
@@ -263,31 +366,116 @@ struct mpc_result SolveMPC(Eigen::Matrix<double, 2, 1> previous_vel, lgmath::se3
...
 
-    
     // Return the resulting structure
     return {applied_vel, mpc_poses};
     }
 }
 
 
+/*
+// Measurement generation overhaul, not quite ready to deploy yet
```
Comment (a-krawciw) [inline main/src/vtr_path_planning/src/mpc/mpc_path_planner.cpp:376]: Does that mean it is coming?

File: main/src/vtr_path_planning/include/vtr_path_planning/cbit/cbit_config.hpp [hunk 6]
```diff
@@ -66,8 +71,6 @@ class CBITConfig {
         double robot_linear_velocity_scale = 1.0;
         double robot_angular_velocity_scale = 1.0;
 
-        // Add unicycle model param
-
         // Covariance Tuning Weights
         //Eigen::Matrix<double, 6, 6> pose_error_cov = Eigen::Matrix<double, 6, 6>::Zero();
```
[the remark this answers is not in the document]
Comment (JordySehn) [reply main/src/vtr_path_planning/include/vtr_path_planning/cbit/cbit_config.hpp:75]: No these can be removed

File: rviz/honeycomb.rviz
Comment (JordySehn) [reply rviz/honeycomb.rviz:1]: I had actually forgotten to do the proper update on the warthog ouster rviz either - Ill go through and update the others too

File: main/src/vtr_tactic/src/tactic.cpp (hunk 3 above)
Comment (JordySehn) [reply main/src/vtr_tactic/src/tactic.cpp:168]: I was finding that this wouldnt trigger the end of path until the robot had localized to the last frame which only occured if the robot had actually moved past the last frame. This meant it basically always ignored the route_completion_threshold check that follows and wasnt reliably ending the repeats when the controller got to the end. I found switching to second to last worked much better.

File: main/src/vtr_path_planning/src/mpc/speed_scheduler.cpp (hunk 4 above)
Comment (JordySehn) [reply main/src/vtr_path_planning/src/mpc/speed_scheduler.cpp:69]: yes it probably should be - will add in next commit

File: main/src/vtr_path_planning/src/mpc/mpc_path_planner.cpp (hunk 5 above)
Comment (JordySehn) [reply main/src/vtr_path_planning/src/mpc/mpc_path_planner.cpp:376]: ooop forgot to remove this - ended up scrapping this early prototype function and just integrated some of the ideas into the existing function - will remove comments

File: main/src/vtr_navigation/src/graph_map_server.cpp [hunk 7]
```diff
@@ -508,6 +508,9 @@ void GraphMapServer::updateVertexType() {
             ->retrieve<tactic::EnvInfo>("env_info",
                                         "vtr_tactic_msgs/msg/EnvInfo");
     vertex.type = env_info_msg->sharedLocked().get().getData().terrain_type;
+    graph->at(VertexId(vertex.id))->SetTerrainType(vertex.type);
```
[the remark this answers is not in the document]
Comment (a-krawciw) [reply main/src/vtr_navigation/src/graph_map_server.cpp]: You need it for CLOG because there is an explicit function header defined for unsigned longs. Without it the conversion is implicitly defined.

I have removed

File: main/src/vtr_pose_graph/include/vtr_pose_graph/path/path.hpp (hunk 2 above)
Comment (a-krawciw) [reply main/src/vtr_pose_graph/include/vtr_pose_graph/path/path.hpp]: Resolved by accessing sequence instead.

File: main/src/vtr_path_planning/src/mpc/speed_scheduler.cpp (hunk 1 above)
Comment (JordySehn) [reply main/src/vtr_path_planning/src/mpc/speed_scheduler.cpp:33]: Removing the TODO's 4 and 5 in next commit. I had implemented these but decided not to add them as during testing I didnt like the effect

**Review by a-krawciw — APPROVED** [review]
